User Mutations

CodeProKid opened the pull request:
This PR adds user mutations. I still have to write tests for this, but I think it's at a good enough point where the code can start getting reviewed.

jasonbahl commented on src/Type/User/Mutation/UserUpdate.php:
```diff
@@ -0,0 +1,132 @@
+					}
+
+					/**
+					 * If the $post_id is empty, we should throw an exception
```
$user_id 😉

jasonbahl commented on src/Type/User/Mutation/UserCreate.php:
```diff
@@ -0,0 +1,118 @@
+		/**
+		 * Create mutations require a login to be passed
+		 */
+		$user_fields['login']['type'] = Types::non_null( Types::string() );
```
since the `UserMutation::input_fields()` is filterable we should probably explicitly make sure the login field exists, similar to how you're doing here: https://github.com/wp-graphql/wp-graphql/pull/191/files#diff-4fdbc70dee8938bb749e25fefa47a18bR114

Let's say someone did a heavy-handed filter like: 

```
add_filter( 'graphql_user_mutation_input_fields', function() {
  return $fields['the_only_mutation_field'] = [
     'type' => Types::string(),
     'description' => __( 'I just filtered out all mutation inputs except this field, bro', 'some-silly-plugin' ),
  ];
}, 10, 1 );
```

This would cause: `$user_fields['login']['type'] = Types::non_null( Types::string() );` to be a PHP warning as the array_key `login` wouldn't exist.

jasonbahl commented on src/Type/User/Mutation/UserDelete.php:
```diff
@@ -0,0 +1,120 @@
+						'description' => __( 'The ID of the user you want to delete', 'wp-graphql' ),
+					],
+					'reassignId' => [
+						'type' => Types::int(),
```
`Types::id()`

jasonbahl commented on src/Type/User/Mutation/UserMutation.php:
```diff
@@ -0,0 +1,209 @@
+			$insert_user_args['user_url'] = esc_url( $input['websiteUrl'] );
+		}
+
+		if ( ! empty( $input['email'] ) ) {
```
turns out that if you enter a plain string instead of an email address for email, the email just gets set to null. . .

So, I'm thinking we need some validation here to ensure the input is an acceptable email address (probably need to dig and see what core is doing to validate the email when the update is made) before we allow it to be used as the `user_email` field for insertion.

I'm thinking, if the email isn't "valid" we just unset it and don't pass it to the mutation at all. . .and/or we throw an exception so the user knows the email was invalid.

See attached GIF. 

![string-emails-not-working](https://user-images.githubusercontent.com/1260765/29326634-92f49d4e-81a9-11e7-8383-9b60e6353c1e.gif)

CodeProKid replied:
I think the exception route would be best for now. Until someone has an objection to doing this, I say we roll with that.

jasonbahl replied:
Ya, let's do some validation on the input to make sure it's an email and throw an exception if it's not proper

CodeProKid commented:
@jasonbahl took a first pass at tests here, still needs some cleanup, and maybe a few more added, but feel free to check it out. Also still need to address some of the previous feedback.

CodeProKid commented on src/Type/User/Mutation/UserCreate.php:
```diff
@@ -0,0 +1,123 @@
+						throw new \Exception( __( 'Mutation not processed. There was no input for the mutation.', 'wp-graphql' ) );
+					}
+
+					if ( ! current_user_can( 'create_users' ) ) {
```
I'll add a followup issue for this, since it's a separate mutation, and will have separate tests.

CodeProKid commented:
@jasonbahl addressed all feedback, and wrote a few tests! I feel pretty good about the coverage now, not sure how to rerun all the tests here though, they are all passing locally for me though.

jasonbahl commented:
looks like tests are barfing for Multisite. . .I'm gonna investigate a bit and see if I can figure out what's up. My guess is user creation, etc is handled a bit different for multisite?

jasonbahl commented:
Found a few things to adjust for multisite. 

Fwiw, to test in multisite locally you can run `export WP_MULTISITE=1` to set your environment variable so PHPUnit will run as a multisite. To switch back to non-multisite testing `export WP_MULTISITE=0`

You can see the travis.yml file as well to see what tests are run, etc. (need to get a good README together RE: unit tests. . .I'm still learning a LOT too).

jasonbahl commented on src/Type/User/Mutation/UserMutation.php:
```diff
@@ -0,0 +1,291 @@
+					'type'        => Types::list_of( Types::string() ),
+					'description' => __( 'An array of roles to be assigned to the user.', 'wp-graphql' ),
+				],
+				'jabber'      => [
```
AIM, YIM and JABBER were removed from core in 3.6 (https://developer.wordpress.org/reference/functions/wp_insert_user/#changelog) so not going to include these in the core user mutations.
